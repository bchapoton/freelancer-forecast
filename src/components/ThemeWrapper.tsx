import React, { ReactNode } from 'react';
import { createTheme, ThemeProvider } from '@mui/material';

export type ThemeWrapperProps = {
    children: ReactNode;
};

function ThemeWrapper({ children }: ThemeWrapperProps) {
    const theme = createTheme({
        typography: {
            fontSize: 13,
        },
        components: {
            MuiTableRow: {
                styleOverrides: {
                    head: {
                        color: '#495057',
                        backgroundColor: '#eff2f7',
                        fontWeight: 600,
                        borderBottomColor: '#d7dade',
                        borderBottomWidth: '2px',
                        borderBottomStyle: 'solid',
                    },
                    hover: {
                        backgroundColor: '#495057',
                    },
                    root: {
                        '&:last-child td, &:last-child th': { border: 0 },
                    },
                },
            },
            MuiTableCell: {
                styleOverrides: {
                    head: {
                        fontSize: '13px',
                        fontWeight: 600,
                        color: '#495057',
                    },
                    body: {
                        color: '#495057',
                        fontSize: '13px',
                    },
                },
            },
            MuiTableContainer: {
                styleOverrides: {
                    root: {
                        borderRadius: 5,
                    },
                },
            },
        },
    });
    return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
}

export default ThemeWrapper;
