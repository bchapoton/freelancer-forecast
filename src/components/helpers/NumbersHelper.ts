export default class NumbersHelper {
    static rangedNumber(value: number, min: number | undefined, max: number | undefined): number {
        let result: number;
        result = max !== undefined && value > max ? max : value;
        result = min !== undefined && value < min ? min : result;
        return result;
    }
}
