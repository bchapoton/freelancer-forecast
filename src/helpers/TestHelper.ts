import { initialState as ParametersInitialState, ParametersState } from '../redux/slices/ParametersSlice';
import { initialState as SummariesInitialState, SummariesState } from '../redux/slices/SummariesSlice';
import { initialState as UIInitialState, UIState } from '../redux/slices/UISlice';

export class TestHelper {
    static initTestingStore(): StorePreloadedState {
        return {
            parameters: {
                ...ParametersInitialState,
                year: 2022,
            },
            summaries: SummariesInitialState,
            ui: UIInitialState,
        };
    }
}

export type StorePreloadedState = {
    parameters: ParametersState;
    summaries: SummariesState;
    ui: UIState;
};
