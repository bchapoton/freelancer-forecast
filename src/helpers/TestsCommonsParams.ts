import { YearSummary } from '../services/ParametersService';
import { ParametersState } from '../redux/slices/ParametersSlice';
import { DAY } from '../services/CalendarService';
import { IncomeTaxesSummary } from '../services/TaxesService';

export const parameters: ParametersState = {
    averageDailyRate: 550,
    year: 2022,
    vatRate: 20,
    taxAllowance: 34,
    socialContribution: [
        { rate: 22.2, label: 'Prestations de services (bnc et bic) et vente de marchandises (bic)' },
        { rate: 0.2, label: 'Formation prof.liberale obligatoire' },
    ],
    taxationFamilyContext: {
        incomeSplittingParts: 3,
        taxableHouseholdRevenues: 24506,
        taxAllowance: 10,
    },
    saving: { mode: 'percentage', value: 20 },
    partTime: { percentage: 100 },
    dayOffContext: { unit: 'weeks', value: 5 },
    nonWorkingDays: [DAY.SATURDAY, DAY.SUNDAY],
};
export const expectedYearSummary: YearSummary = {
    monthSummaries: [
        {
            month: 0,
            totalDays: 19,
            revenue: 10450,
            vat: 2090,
            includingTax: 12540,
            socialContribution: 2340.7999999999997,
            profits: 8109.200000000001,
        },
        {
            month: 1,
            totalDays: 18,
            revenue: 9900,
            vat: 1980,
            includingTax: 11880,
            socialContribution: 2217.6,
            profits: 7682.4,
        },
        {
            month: 2,
            totalDays: 21,
            revenue: 11550,
            vat: 2310,
            includingTax: 13860,
            socialContribution: 2587.2,
            profits: 8962.8,
        },
        {
            month: 3,
            totalDays: 19,
            revenue: 10450,
            vat: 2090,
            includingTax: 12540,
            socialContribution: 2340.7999999999997,
            profits: 8109.200000000001,
        },
        {
            month: 4,
            totalDays: 20,
            revenue: 11000,
            vat: 2200,
            includingTax: 13200,
            socialContribution: 2463.9999999999995,
            profits: 8536,
        },
        {
            month: 5,
            totalDays: 20,
            revenue: 11000,
            vat: 2200,
            includingTax: 13200,
            socialContribution: 2463.9999999999995,
            profits: 8536,
        },
        {
            month: 6,
            totalDays: 19,
            revenue: 10450,
            vat: 2090,
            includingTax: 12540,
            socialContribution: 2340.7999999999997,
            profits: 8109.200000000001,
        },
        {
            month: 7,
            totalDays: 21,
            revenue: 11550,
            vat: 2310,
            includingTax: 13860,
            socialContribution: 2587.2,
            profits: 8962.8,
        },
        {
            month: 8,
            totalDays: 20,
            revenue: 11000,
            vat: 2200,
            includingTax: 13200,
            socialContribution: 2463.9999999999995,
            profits: 8536,
        },
        {
            month: 9,
            totalDays: 19,
            revenue: 10450,
            vat: 2090,
            includingTax: 12540,
            socialContribution: 2340.7999999999997,
            profits: 8109.200000000001,
        },
        {
            month: 10,
            totalDays: 20,
            revenue: 11000,
            vat: 2200,
            includingTax: 13200,
            socialContribution: 2463.9999999999995,
            profits: 8536,
        },
        {
            month: 11,
            totalDays: 19,
            revenue: 10450,
            vat: 2090,
            includingTax: 12540,
            socialContribution: 2340.7999999999997,
            profits: 8109.200000000001,
        },
    ],
    totals: {
        totalDays: 235,
        includingTax: 155100,
        socialContribution: 28952,
        vat: 25850,
        profits: 100298,
        revenue: 129250,
    },
};
export const expectedIncomeTaxesSummaryAlone: IncomeTaxesSummary = {
    taxBracketSummaries: [
        { bracket: { bottomBracket: 0, topBracket: 10225, rate: 0 }, amount: 0 },
        { bracket: { bottomBracket: 10226, topBracket: 26070, rate: 11 }, amount: 1742.84 },
        { bracket: { bottomBracket: 26071, topBracket: 74545, rate: 30 }, amount: 11978.699999999995 },
        { bracket: { bottomBracket: 74546, topBracket: 160336, rate: 41 }, amount: 0 },
        { bracket: { bottomBracket: 160336, rate: 45 }, amount: 0 },
    ],
    netTaxableIncome: {
        contractorNetTaxableIncome: 65999.99999999999,
        familyNetTaxableIncome: 0,
        incomeSplittingParts: 1,
        netTaxableIncomeResult: 65999.99999999999,
    },
    total: 13721.539999999995,
};
