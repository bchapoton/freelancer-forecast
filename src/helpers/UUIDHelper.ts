import { v4 } from 'uuid';

export function generateUID(): string {
    return v4();
}
