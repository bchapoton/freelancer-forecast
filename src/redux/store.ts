import { configureStore } from '@reduxjs/toolkit';
import parametersReducer from './slices/ParametersSlice';
import summariesReducer from './slices/SummariesSlice';
import uiReducer from './slices/UISlice';
import { StorePreloadedState } from '../helpers/TestHelper';

export const store = storeConfigurator();

export function storeConfigurator(preloadedState?: StorePreloadedState) {
    return configureStore({
        reducer: {
            parameters: parametersReducer,
            summaries: summariesReducer,
            ui: uiReducer,
        },
        preloadedState: preloadedState,
    });
}

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
