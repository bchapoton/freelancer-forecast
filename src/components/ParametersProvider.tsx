import React, { Fragment, ReactNode, useEffect } from 'react';
import { useAppDispatch, useAppSelector } from '../redux/hooks';
import { ParametersState, selectParameters } from '../redux/slices/ParametersSlice';
import parametersService, { YearSummary } from '../services/ParametersService';
import { setIncomeTaxesSummary, setYearSummary } from '../redux/slices/SummariesSlice';
import taxesService, { IncomeTaxesSummary } from '../services/TaxesService';

export type ParametersProviderProps = {
    children: ReactNode;
};

function ParametersProvider({ children }: ParametersProviderProps) {
    const parameters: ParametersState = useAppSelector(selectParameters);
    const dispatch = useAppDispatch();

    useEffect(() => {
        const yearSummary: YearSummary = parametersService.buildYearSummary(parameters);
        dispatch(setYearSummary(yearSummary));
        const incomeTaxesSummary: IncomeTaxesSummary = taxesService.getIncomeTaxesSummary(
            yearSummary.totals.revenue,
            parameters.taxAllowance,
            parameters.taxationFamilyContext,
        );
        dispatch(setIncomeTaxesSummary(incomeTaxesSummary));
    }, [dispatch, parameters]);

    return <Fragment>{children}</Fragment>;
}

export default ParametersProvider;
