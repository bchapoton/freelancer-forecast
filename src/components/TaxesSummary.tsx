import React, { Fragment } from 'react';
import {
    Box,
    Paper,
    Table,
    TableBody,
    TableCell,
    TableContainer,
    TableHead,
    TableRow,
    Typography,
} from '@mui/material';
import { IncomeTaxesSummary, TaxBracketSummary } from '../services/TaxesService';
import Percentage from './Percentage';
import Euro from './Euro';
import { YearSummary } from '../services/ParametersService';
import { useAppSelector } from '../redux/hooks';
import { selectIncomeTaxesSummary, selectYearSummary } from '../redux/slices/FinancialSlice';
import { ParametersState, selectParameters } from '../redux/slices/ParametersSlice';
import { TableCellBold, TableRowFooter } from './TableCustomComponents';
import { BoxedParameters } from './ParametersSummary';

function TaxesSummary() {
    const yearSummary: YearSummary = useAppSelector(selectYearSummary);
    const parameters: ParametersState = useAppSelector(selectParameters);
    const incomeTaxesSummary: IncomeTaxesSummary = useAppSelector(selectIncomeTaxesSummary);

    return (
        <Fragment>
            <Typography gutterBottom variant="h5" component="div">
                Imp√¥ts sur le revenu
            </Typography>
            <Paper>
                <Box sx={{ p: 2 }}>
                    <BoxedParameters
                        noWrap={false}
                        title="Revenus entreprise ( net imposable )"
                        value={
                            <Fragment>
                                <Euro>{yearSummary.totals.revenue}</Euro>&nbsp;(&nbsp;
                                <Euro>{incomeTaxesSummary.netTaxableIncome.contractorNetTaxableIncome}</Euro>&nbsp;)
                            </Fragment>
                        }
                    />
                    <BoxedParameters
                        noWrap={false}
                        title="Revenus famillaux ( net imposable )"
                        value={
                            <Fragment>
                                <Euro>{parameters.taxationFamilyContext.taxableHouseholdRevenues}</Euro>&nbsp;(&nbsp;
                                <Euro>{incomeTaxesSummary.netTaxableIncome.familyNetTaxableIncome}</Euro>
                                &nbsp;)
                            </Fragment>
                        }
                    />
                    <BoxedParameters
                        noWrap={false}
                        title="Nombre de parts"
                        value={incomeTaxesSummary.netTaxableIncome.incomeSplittingParts + ' parts'}
                    />
                    <BoxedParameters
                        noWrap={false}
                        highlighted
                        title="Revenu net imposable"
                        value={<Euro>{incomeTaxesSummary.netTaxableIncome.netTaxableIncomeResult}</Euro>}
                    />
                </Box>
                <TableContainer>
                    <Table size="small">
                        <TableHead>
                            <TableRow>
                                <TableCell>Tranche imposition</TableCell>
                                <TableCell align="right">Taux imposition</TableCell>
                                <TableCell align="right">montant par tranche</TableCell>
                            </TableRow>
                        </TableHead>
                        <TableBody>
                            {incomeTaxesSummary.taxBracketSummaries.map((taxBracketResult: TaxBracketSummary) => (
                                <TableRow
                                    key={
                                        taxBracketResult.bracket.bottomBracket +
                                        '-' +
                                        taxBracketResult.bracket.topBracket
                                    }
                                    hover={true}
                                >
                                    <TableCell component="th" scope="row">
                                        <Euro cents={false}>{taxBracketResult.bracket.bottomBracket}</Euro>
                                        <span> - </span>
                                        <Euro cents={false}>{taxBracketResult.bracket.topBracket}</Euro>
                                    </TableCell>
                                    <TableCell align="right">
                                        <Percentage>{taxBracketResult.bracket.rate}</Percentage>
                                    </TableCell>
                                    <TableCell align="right">
                                        <Euro>{taxBracketResult.amount}</Euro>
                                    </TableCell>
                                </TableRow>
                            ))}
                            <TableRowFooter>
                                <TableCellBold component="th" scope="row">
                                    Total
                                </TableCellBold>
                                <TableCellBold align="right" colSpan={2}>
                                    <Euro>{incomeTaxesSummary.total}</Euro>
                                </TableCellBold>
                            </TableRowFooter>
                        </TableBody>
                    </Table>
                </TableContainer>
            </Paper>
        </Fragment>
    );
}

export default TaxesSummary;
