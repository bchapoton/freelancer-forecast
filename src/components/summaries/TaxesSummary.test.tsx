import React from 'react';
import renderer from 'react-test-renderer';
import { storeConfigurator } from '../../redux/store';
import { Provider } from 'react-redux';
import ThemeWrapper from '../ThemeWrapper';
import TaxesSummary from './TaxesSummary';
import { TestHelper } from '../../helpers/TestHelper';
import ParametersProvider from '../ParametersProvider';

it('Taxes summary renders correctly', () => {
    const tree = renderer
        .create(
            <Provider store={storeConfigurator(TestHelper.initTestingStore())}>
                <ThemeWrapper>
                    <ParametersProvider>
                        <TaxesSummary />
                    </ParametersProvider>
                </ThemeWrapper>
            </Provider>,
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});
