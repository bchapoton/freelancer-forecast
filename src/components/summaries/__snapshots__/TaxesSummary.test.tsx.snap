// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Taxes summary renders correctly 1`] = `
Array [
  <div
    className="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-1pzmhef-MuiTypography-root"
  >
    Impôts sur le revenu
  </div>,
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1ps6pg7-MuiPaper-root"
  >
    <div
      className="MuiBox-root css-19midj6"
    >
      <div
        className="MuiBox-root css-k008qs"
      >
        <div
          className="MuiTypography-root MuiTypography-subtitle2 css-z1jfgo-MuiTypography-root"
        >
          Revenus entreprise (net imposable)
        </div>
        <div
          className="MuiTypography-root MuiTypography-body2 css-1umiv4o-MuiTypography-root"
        >
          <span
            className="MuiBox-root css-1l7mgse"
          >
            0,00 €
          </span>
           (
          <span
            className="MuiBox-root css-1l7mgse"
          >
            0,00 €
          </span>
          )
        </div>
      </div>
      <div
        className="MuiBox-root css-k008qs"
      >
        <div
          className="MuiTypography-root MuiTypography-subtitle2 css-z1jfgo-MuiTypography-root"
        >
          Revenus famillaux ( net imposable )
        </div>
        <div
          className="MuiTypography-root MuiTypography-body2 css-1umiv4o-MuiTypography-root"
        >
          <span
            className="MuiBox-root css-1l7mgse"
          >
            24 506,00 €
          </span>
           (
          <span
            className="MuiBox-root css-1l7mgse"
          >
            0,00 €
          </span>
          )
        </div>
      </div>
      <div
        className="MuiBox-root css-k008qs"
      >
        <div
          className="MuiTypography-root MuiTypography-subtitle2 css-z1jfgo-MuiTypography-root"
        >
          Nombre de parts
        </div>
        <div
          className="MuiTypography-root MuiTypography-body2 css-1umiv4o-MuiTypography-root"
        >
          1 parts
        </div>
      </div>
      <div
        className="MuiBox-root css-14cn5a1"
      >
        <div
          className="MuiTypography-root MuiTypography-subtitle2 css-z1jfgo-MuiTypography-root"
        >
          Revenu net imposable
        </div>
        <div
          className="MuiTypography-root MuiTypography-body2 css-1umiv4o-MuiTypography-root"
        >
          <span
            className="MuiBox-root css-1l7mgse"
          >
            0,00 €
          </span>
        </div>
      </div>
    </div>
    <div
      className="MuiTableContainer-root css-1lqbn85-MuiTableContainer-root"
    >
      <table
        className="MuiTable-root css-1m14wjh-MuiTable-root"
        role={null}
      >
        <thead
          className="MuiTableHead-root css-15wwp11-MuiTableHead-root"
          role={null}
        >
          <tr
            className="MuiTableRow-root MuiTableRow-head css-xev7fm-MuiTableRow-root"
            role={null}
          >
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-t8mu5n-MuiTableCell-root"
              scope="col"
            >
              Tranche imposition
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeSmall css-abrxy2-MuiTableCell-root"
              scope="col"
            >
              Taux imposition
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeSmall css-abrxy2-MuiTableCell-root"
              scope="col"
            >
              montant par tranche
            </th>
          </tr>
        </thead>
        <tbody
          className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
          role={null}
        >
          <tr
            className="MuiTableRow-root css-15p7zew-MuiTableRow-root"
            role={null}
          >
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1off7ch-MuiTableCell-root"
              scope="row"
            >
              Total
            </th>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall css-ftv5mp-MuiTableCell-root"
              colSpan={2}
            >
              <span
                className="MuiBox-root css-1l7mgse"
              >
                0,00 €
              </span>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>,
]
`;
